### ISSUE TITLE
Obtaining atomic orbital quantities

### ISSUE THREAD

**Issue by sangyeon-hwang** [orbkit/orbkit#10]
Hello.
I'm currently working on obtaining densities and derivatives of AOs (rather than MOs), mainly using Gaussian fchk files.
I have some related questions.
(I have a version where the last commit date is Sep 7, 2017.)

**1. Error in AO calculation**
When I run the following code, testing the AO values on an arbitrary point:
```python
qc = orbkit.read.main_read('c.fchk', 'gaussian.fchk')
orbkit.options.numproc = 10
orbkit.options.grid_file = 'points'  # Has coordinate of a point.
orbkit.grid.is_vector = True
orbkit.options.calc_ao = True
data = orbkit.run_orbkit(qc)
```
an error occurs as
```
  File "<orbkit root>/extras.py", line 415, in get_ao
    global_args['ao_spec'][x]['exp_list'][0],
TypeError: only integer scalar arrays can be converted to a scalar index
```
`c.fchk` is a test fchk file of a carbon atom with HF/6-31G\*.
I think lines 415-416 of `extras.py` should be changed from (omitting indentation)
```
global_args['ao_spec'][x]['exp_list'][0],
global_args['ao_spec'][x]['atom'])
```
to
```
global_args['ao_spec'][int(x)]['exp_list'][0],
global_args['ao_spec'][int(x)]['atom'])
```

**2. SP shell handling in pure harmonic case**
It seems `ao_spherical` creation in `read.read_gaussian_fchk` is incomplete when the basis is pure harmonic and has SP shell types.
For example, if the shell types of my fchk are `[0, -1, ...]`, I suppose `ao_spherical` should be
```
[[0, (0, 0)], [1, (0, 0)], [2, (1, 1)], [2, (1, -1)], [2, (1, 0)], ...]
```
yet in actual test run, I get
```
[[0, (0, 0)], [1, (1, 1)], [1, (1, -1)], [1, (1, 0)], ...]
```
This causes the number of AOs to be deficient when I use `core.cartesian2spherical` to match the AO size to `Number of basis functions` of fchk.

**3. Performance difference of AO/MO routines**
As `core.ao_creator` seems to run only serially and accept only up to 1st-order derivatives, I'm mostly utilizing the high-level interface.
While doing so, the following test run:
```python
qc = orbkit.read.main_read('thio.fchk', 'gaussian.fchk')  # Basis size = 278
orbkit.options.numproc = 10
grid = numpy.random.rand(2_000_000, 3)
orbkit.grid.x = grid[:,0]
orbkit.grid.y = grid[:,1]
orbkit.grid.z = grid[:,2]
orbkit.grid.is_initialized = True
orbkit.grid.is_vector = True
orbkit.options.calc_mo = [43]
orbkit.options.drv = ['None', 'x', 'y', 'z', 'xx', 'yy', 'zz']
data = orbkit.run_orbkit(qc)
```
with 10 cores of Intel Xeon E5 ended after 25 secs, yet the other run:
```python
## Same qc and grid setting
orbkit.options.calc_ao = True
orbkit.options.drv = ['None', 'x', 'y', 'z', 'xx', 'yy', 'zz']
data = orbkit.run_orbkit(qc)
```
ended after 72 secs using 228-fold RAM of the former.
Could I know the origin of this performance difference?
(I can understand the increase of memory use, concerning the sizes of the returns.)
To interface with my code, it will be much more helpful to obtain AO than MO values, so I want to check my utilization of `orbkit` is appropriate.

Thank you for your reading.

**Comment by sangyeon-hwang** [comment]
About Question 2, I tried changing lines 1074-1086 of `read.py` as follows:
```python
  # Look for SP atomic orbitals
  if ao_sp_coeffs:
    ao_new = [] 
    for i,ao in enumerate(qc.ao_spec):
      if ao['type'] == 'p' and sum(numpy.abs(ao_sp_coeffs[i])) > 0: 
        ao_new.append(copy.deepcopy(ao))
        ao_new[-1]['type'] = 's'
        ao_new.append(ao)
        ao_new[-1]['type'] = 'p'
        ao_new[-1]['coeffs'][:,1] = numpy.array(ao_sp_coeffs[i])     
      else:
        ao_new.append(ao)
    ## NEW LINES BEGIN.
    #+ Correct `qc.ao_spherical` for pure SP shells.
    if not cartesian_basis:
      ao_spherical_new = [] 
      count = 0
      for i,ao in enumerate(qc.ao_spec):
        if ao['type'] == 'p' and sum(numpy.abs(ao_sp_coeffs[i])) > 0: 
          ao_spherical_new.append([count,(0,0)])
          count += 1 
          for m in (1, -1, 0):
            ao_spherical_new.append([count,(1,m)])
          count += 1 
        else:
          l = lquant[ao['type']]
          for m in (range(0,l+1) if l != 1 else [1,0]):
            ao_spherical_new.append([count,(l,m)])
            if m != 0:
              ao_spherical_new.append([count,(l,-m)])
          count += 1
      ## A simple check
      if count != len(ao_new):
        raise RuntimeError("Incorrect 'ao_spherical' creation")
      qc.ao_spherical = ao_spherical_new
    ## NEW LINES END.
    qc.ao_spec = ao_new   
```
By running a few tests I checked results come correctly.
I wonder my modification is okay with other routines.

**Comment by vpohl** [comment]
Dear Sang-Yeon Hwang,

thank you very much for reporting these issues. I will look into it as soon as possible. Can you maybe send me your test fchk file that I can reproduce the first two issues?

With best regards,
Vincent

**Comment by sangyeon-hwang** [comment]
Dear Vincent Pohl:

Thank you for your reply.
I here attach my fchk file. I edited the extension due to the git policy.
[c.fchk.txt](https://github.com/orbkit/orbkit/files/1316205/c.fchk.txt)

Sincerely, Sang-Yeon.

**Comment by vpohl** [comment]
Dear Sang-Yeon,

I applied and commited your bugfix. Further, I improved the calc_ao interface. It now uses the same framework as all other grid-based routines (rho_compute). Thus, it paralellizes over the grid and supports all available derivatives (x, xx, xy, etc.).

Can you please test the interface ([https://github.com/orbkit/orbkit/tree/improve_calc_ao](url)), before I merge it into the cython branch?

Thank you very much!

With best regards,
Vincent

**Comment by sangyeon-hwang** [comment]
Hello, Vincent.

I've tested the code of the new branch.
I see the mentioned problems have been corrected and values come out as expected.
Thank you for improving the AO interface. It was already fast actually, yet it became certainly better.
It seems I do not need to use `core.cartesian2spherical` any more as now the AO data dimension automatically matches the actual basis size in pure harmonic cases. This was also a pleasant change!

By the way, I had to fix two minor things in the new code.
There was a syntax error in lines 743-5 of `core.py`:
```python
  display('\nThere are %d contracted %s AOs' % (len(Spec['mo_spec'][0]['coeffs']),
          'Cartesian' if not Spec['ao_spherical'] else 'spherical')+ 
          ('' if calc_ao else ' and %d MOs to be calculated.' % len(Spec['mo_spec']) )
```
I think there should be another closing parenthesis at the end of the last line.

Also, about line 280 of `analytical_integrals.py`:
```python
  ij = zip(ij[:-1],ij[1:])
```
as I write in Python 3.6, I needed to enclose the right hand with `list()`. Without it, an error occurs as
```
  File "<orbkit root>/analytical_integrals.py", line 294, in get_mo_overlap_matrix
    mo_overlap_matrix[m:n,:] = it.next() if numproc > 1 else get_slice(ij[l])
TypeError: 'zip' object is not subscriptable
```
when I try to use `analytical_integrals.get_mo_overlap_matrix`.

Other than those above, everything was fine, at least from my current uses.
Thank you for the fixes!

Sincerely, Sang-Yeon.
